src/steamship/plugin/capabilities.py

```diff
@@ -0,0 +1,291 @@
+        return CapabilityPluginResponse(capability_responses=responses)
+
+    def extract_from_blocks(self, blocks: Iterable[Block]) -> CapabilityPluginResponse:
+        """Find the block in a list that defines capability requests, and initialize this data structure with it.
+
+        It may be the case that there is no block indicating capability requests; Older clients may be passing blocks
+        to the plugin without programmatic knowledge of capabilities.
+
+        :param blocks: A list of blocks that was passed as input to a plugin
+        :return: a CapabilityPluginResponse which indicates the level at which we are serving each request
+        """
+        capabilities_block = None
+        for block in blocks:
+            if block.mime_type == MimeTypes.STEAMSHIP_PLUGIN_CAPABILITIES:
+                if capabilities_block is not None:
+                    logging.error(
+                        f"Found more than one block with MIME_TYPE {MimeTypes.STEAMSHIP_PLUGIN_CAPABILITIES} in request blocks.  Using first one found."
+                    )
+                    break
+                capabilities_block = block
+        if not capabilities_block:
+            return CapabilityPluginResponse(capability_responses=[])
+        return self.load_requests(CapabilityPluginRequest.parse_raw(capabilities_block.text))
+
+    def get(
+        self, typ: Type[CapabilityType], default: Optional[CapabilityType] = None
+    ) -> Optional[CapabilityType]:
+        if self._requested is None:
+            raise SteamshipError(
+                "RequestedCapabilities has not been loaded with a set of requests yet.  Load a request with .load_requests()"
+            )
+        capability = self._requested.get(typ.NAME)
+        if capability:
+            return typ.parse_obj(capability)
+        return default
+
+
+class SystemPromptSupport(CapabilityImpl):
+    """This plugin supports system prompts separate from the per-request model prompt.
+
+    The system prompt will come across in other blocks on the request.
+    """
+
+    name = "steamship.system_prompt_support"
+
+    request_level = RequestLevel.BEST_EFFORT
+    """
+    If NATIVE, asserts that the model being used supports System Prompt in a first-class capacity.
+    If BEST_EFFORT, the system prompt will be incorporated into the request in another way, usually via concatenation
+      with the prompt.
+    """
+
+
+class ConversationSupport(CapabilityImpl):
+    """This plugin supports conversations.
+
+    The content of the conversation will come across in other blocks on the request, using the CHAT TagKind.
+    """
+
+    name = "steamship.conversation_support"
+
+
+class FunctionCallingSupport(CapabilityImpl):
+    """This plugin supports function calling.
+
+    Function definitions come across as a list of function definitions.
+    """
+
+    name = "steamship.function_calling_support"
+
+    functions: List[OpenAIFunction]
+    # TODO (PR): Worth generalizing this?
```

Comment:
Discussion: at very least it seems like generalizing the name here makes sense.

Reply:
I think this should probably be `Tool`-specific naming, just to keep things clearly in Steamship-land and not OpenAI land.

I was hoping we'd have some serialization of `Tool`, such that this could be a `List[Tool]`.

One thing we'll want is some way to specify the serialization of params for a `Tool`. Right now, we built it assuming everything is blocks-in, blocks-out. But our users don't want to wrap the parsing internally.

So, having some meta wrapper that allows `List[Block] --> params` would be extra credit.

Reply:
I'd much prefer that we serialize `Tool` and pass that around. `List[Tool]`. That would prevent any sort of intermediate step in the SDK that is backend specific (until OpenAIFunction becomes a _de facto_ standard, at least).

Bonus points if we can figure out a way to include parameterization of `Tool::run` in a way that will auto-handle `Block --> param` translation, so that devs can write code like `def run(self, query: str, location: str)` and not have to worry about extracting those values from `List[Block]`.